### DISCUSSION TITLE
[vcpkg-msbuild-integration] Extending logic that decides between 'Release' and 'Debug' libraries by referring to $(UseDebugLibraries)

### DISCUSSION THREAD

**Discussion by trueqbit** [microsoft/vcpkg#19144] [Ideas]
****

**Problem space**
Multiple people already tweaked the way _vcpkg.targets_ determines whether to deal with 'Release' or 'Debug' builds for MSBuild ([@paercebal](https://github.com/microsoft/vcpkg/commit/f1c6b5cd4b581ec8d3a1b86fe01c4444e3906b2d), [@ras0219-msft](https://github.com/microsoft/vcpkg/commit/330b8d8bab6a3d07165bf7c05fea09a8e0d56348)).

MSBuild features a build variable called `UseDebugLibraries`, which is used by MSBuild itself to decide between 'Release' and 'Debug' builds and is IMO the most accurate way to do so instead of checking naming conventions.
Also, this would save me from renaming Visual C++ configurations whose names don't start with "Release" or start with "Debug" but are variants using the 'Release' CRT.

**Proposed solution**
_scripts/buildsystems/msbuild/vcpkg.targets_ should be extended to determine the build configuration in the following way:

```xml
<VcpkgNormalizedConfiguration Condition="'$(UseDebugLibraries)' == 'true'">Debug</VcpkgNormalizedConfiguration>
<VcpkgNormalizedConfiguration Condition="'$(UseDebugLibraries)' == 'false'">Release</VcpkgNormalizedConfiguration>
<VcpkgNormalizedConfiguration Condition="'$(VcpkgNormalizedConfiguration)' == '' and $(VcpkgConfiguration.StartsWith('Debug'))">Debug</VcpkgNormalizedConfiguration>
<VcpkgNormalizedConfiguration Condition="'$(VcpkgNormalizedConfiguration)' == '' and ($(VcpkgConfiguration.StartsWith('Release')) or '$(VcpkgConfiguration)' == 'RelWithDebInfo' or '$(VcpkgConfiguration)' == 'MinSizeRel')">Release</VcpkgNormalizedConfiguration>
```

**Comment by BillyONeal** [comment]
That seems a reasonable solution to me

**Comment by BillyONeal** [comment]
(Although Robert made relevant comments in the PR)

**Comment by PhoebeHui** [comment]
Related to similar issue https://github.com/microsoft/vcpkg/issues/2940

**Comment by trueqbit** [comment]
Hello everybody,

my initial PR #9496 was flagged "requires:author-response". Thank you for your efforts in regard to triplets and improving the Msbuild buildsystem integration!
Since I can't reopen a PR, I opened a sequel with #22086.

With the advent of `VcpkgConfiguration` in [vcpkg.props](https://github.com/microsoft/vcpkg/commits/14514508d8d30bdbd645b2bec89696aec25497f1/scripts/buildsystems/msbuild/vcpkg.props), developers can now at least override how vcpkg sees the current configuration.

However, this still feels unnatural and cumbersome, as I have to update all projects to set `<VcpkgConfiguration Condition="'$(Configuration)'=='DebugRelease'">Release</VcpkgConfiguration>`.

Ultimately, the current logic is just an attempt to make the most of accepted naming conventions (from cmake).
I would still hope for a good sensible default of a normalized configuration.

In this sense I presume that someone from the Visual Studio/C++/Msbuild team has had the same thoughts on this when introducing the UseDebugLibraries property.

I underline my point by how UseDebugLibraries is used by Visual C++ and Msbuild itself.

- The Visual C++ project wizard explicitly sets UseDebugLibraries in the "Configuration" property group (![project configuration screenshot](https://user-images.githubusercontent.com/5430212/147109922-21d32ceb-8a5f-4456-af44-a12c406243bf.png)).
- UseDebugLibraries is used in 45 places in various .props/.targets files coming with Msbuild ([UseDebugLibraries occurrences.txt](https://github.com/microsoft/vcpkg/files/7763263/UseDebugLibraries.occurrences.txt)).
- Among other various choices for compiler and linker flags, UseDebugLibraries governs the selection of the runtime, both in _Microsoft.Cl.Common.props_ and _Microsoft.Cpp.mfc*.props_, or the definition of _DEBUG.
